nano-sync/src/gadgets/pedersen.rs

```diff
@@ -0,0 +1,140 @@
+use std::marker::PhantomData;
+
+use algebra_core::{PrimeField, SWModelParameters};
+use r1cs_core::SynthesisError;
+use r1cs_std::boolean::Boolean;
+use r1cs_std::fields::FieldGadget;
+use r1cs_std::groups::curves::short_weierstrass::AffineGadget;
+use r1cs_std::groups::GroupGadget;
+use r1cs_std::prelude::CondSelectGadget;
+
+/// This is a gadget that calculates a Pedersen hash. It is collision resistant, but it's not
+/// pseudo-random. Furthermore, its input must have a fixed-length. The main advantage is that it
+/// is purely algebraic and its output is an elliptic curve point.
+/// The Pedersen hash guarantees that the exponent of the resulting point H is not known, the same
+/// can't be said of the Pedersen commitment. Also, note that the Pedersen hash is collision-resistant
+/// but it is not pseudo-random.
+pub struct PedersenHashGadget<
+    P: SWModelParameters,
+    ConstraintF: PrimeField,
+    F: FieldGadget<P::BaseField, ConstraintF>,
+> {
+    _params: PhantomData<P>,
+    _cfield: PhantomData<ConstraintF>,
+    _field: PhantomData<F>,
+}
+
+impl<P: SWModelParameters, ConstraintF: PrimeField, F: FieldGadget<P::BaseField, ConstraintF>>
+    PedersenHashGadget<P, ConstraintF, F>
+{
+    /// Calculates the Pedersen hash. Given a vector of generators G_i and a vector of bits b_i, the
+    /// hash is calculated like so:
+    /// H = G_0 + b_1 * G_1 + b_2 * G_2 + ... + b_n * G_n
+    /// where G_0 is a sum generator that is used to avoid that the sum starts at zero (which is
+    /// problematic because the circuit can't handle addition with zero).
+    pub fn evaluate<CS: r1cs_core::ConstraintSystem<ConstraintF>>(
+        mut cs: CS,
+        input: &Vec<Boolean>,
+        generators: &Vec<AffineGadget<P, ConstraintF, F>>,
```

Comment:
Yes, it is only one bit per generator.
Well, the literature on Pedersen hashes is confusing to say the least.
For Pedersen commitments it is known that you can commit to a message m of the same size as the group order (for us it would be 752 bits if I'm not mistaken) and it is collision-resistant (after all, it is a commitment!). We also have Pedersen vector commitments where we have one generator for each message. In fact, normal Pedersen commitments already work like that, we are also committing to a randomness using a second generator.

Reply:
That reminds that I think I used Pedersen commitments for the Merkle tree, but I'm not sure if it is safe or if we need to use Pedersen hashes...

Reply:
After discussing this on the ZEXE Telegram channel, there is in fact no need to have different functions for this. We can let the Pedersen hash have each generator use all of its capacity and we can use it in the Merkle Trees.

Reply:
Done!